File: easy/brace-matching/.solution/brace-matching.cpp
#include <stack>
#include <string>

bool brace_matching(std::string string) {
    std::stack<char> bracket_stack;
    for(auto it = string.begin(); it != string.end(); ++it) {
        char const& c = *it;
        switch (c)
        {
        case '(':
        case '[':
        case '{':
            bracket_stack.push(c);
            break;
        case '\'':
        case '"':
            if(it == string.begin() || *(it - 1) != '\\') {
                if(c == bracket_stack.top()) {
                    bracket_stack.pop();
                } else {
                    bracket_stack.push(c);
                }
            }
            break;
        case ')':
        case ']':
        case '}':
            if(c != bracket_stack.top()) {
                return false;
            } else {
                bracket_stack.pop();
            }
            break;
        }
    }
    return bracket_stack.empty();
}

File: easy/brace-matching/.solution/BraceMatching.java
import java.util.Stack;

public class BraceMatching {
    private static boolean isMatchingPair(char opening, char closing) {
        return (opening == '(' && closing == ')') ||
               (opening == '{' && closing == '}') ||
               (opening == '[' && closing == ']');
    }

    public static boolean isBalanced(String s) {
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else if (c == ')' || c == '}' || c == ']') {
                if (stack.isEmpty())
                    return false;
                char top = stack.pop();
                if (!isMatchingPair(top, c))
                    return false;
            }
        }
        return stack.isEmpty();
    }
    
    private static void test(String input) {
        System.out.print("\"" + input + "\": ");
        if (isBalanced(input)) {
            System.out.println("Balanced");
        } else {
            System.out.println("Not Balanced");
        }
    }

    public static void main(String[] args) {
        test("");
        test("(");
        test("()");
        test("(]");
        test("{[}]");
        test("{[()]}");
        test("{[(])}");
        test("{[]()}");
    }
}